docs/vcf_to_bigquery.md

```diff
@@ -0,0 +1,134 @@
+Avro files have
+[many advantages](https://cloud.google.com/bigquery/docs/loading-data-cloud-storage-avro#advantages_of_avro)
+for this purpose, but for our use case, the main advantage is fast and
+[free loading](https://cloud.google.com/bigquery/pricing#free)
+of Avro files into BigQuery. You can follow the progress of this stage from your
+Google Cloud console page:
+
+![Avro to BigQuery](images/vcf_to_bq_load_avro.png)
+
+BigQuery tables generated at this stage are optimized to answer *variant lookup*
+queries such as:
+
+```
+Find all variants in a genomic region
+``` 
+
+That's because we [partition](sharding.md) these tables based on the `start_position` column.
+For this reason, we call this set of tables *variant-optimized* tables.
+
+Some users, as a part of their workflow, need to run *sample lookup* queries such as:
+
+```
+Find all variant of sample X
+```
+
+Executing this type of query on variant-optimized tables is costly, 
+because all rows need to be processed. The cost of such queries can be significant,
+especially if sample lookup queries are executed frequently. For cases like this,
+we offer another set of tables called *sample-optimized* tables. As their name suggests,
+they are optimized for answering queries of the latter type. Sample-optimized tables are
+generated in stage 3 of VCF to BigQuery. These tables are not generated by default;
+users need to set a special flag to activate this functionality.
+
+## Stage3: Create Sample Optimized Tables
+
+Users can store a sample-optimized table by setting their base name using the following flag:
+
+```
+--sample_lookup_optimized_output_table GOOGLE_CLOUD_PROJECT:BIGQUERY_DATASET.SAMPLE_OPT_TABLE
+```
+
+Setting this flag will cause a second set of tables to be generated. Similar to variant-optimized
+tables, these tables will be [sharded](sharding.md) and have the same suffixes (extracted from 
+sharding config file).
+
+
+Storing sample-optimized tables will increase
+your BigQuery [storage costs](https://cloud.google.com/bigquery/pricing#pricing_summary),
+but the extra cost will be quickly recovered through *significant* saving of reduced cost of sample
+lookup queries.
+
+Sample-optimized tables have a schema which is very similar to the original variant-optimized
+tables. There are only two differences:
+
+* Repeated call column is flattened, so that each call ends up in a separate row.
+* A new `sample_id` column, which has the same value as `call.sample_id column`, is added.
+
+Let's consider a concrete example: variant-optimized tables of [phase 3 1000Genome](https://www.internationalgenome.org/data/)
+dataset have `6,468,094` rows in its `__chr1` shard. We know phase 3
+includes `2504` individuals, so each row of this table has `2504` repeated `call` fields.
+If we execute a sample lookup query on this table, it will process the whole table,
+which is `408.13 GB`. According to the current [BigQuery pricing model](https://cloud.google.com/bigquery/pricing#queries),
```

Comment:
If we have a storage price for this table, lets include it too. I presume it's going to be around 4$/mo?

Reply:
If you do that then the break-even point is 11 instead of 27.

![query_rounds_to_break_even](https://user-images.githubusercontent.com/6555937/86389810-ec928c00-bc64-11ea-93de-2382c3c0b2d7.png)

Reply:
The document says the breakpoint ***in queries*** is at least 14, not 27 - 27 is the monthly cost for the sample-optimized table, with query cost being almost negligible, so breaking point ***in dollars*** is 27 plus change.

The doc describes definitive upper bound (hence _at least_), neglecting the addition of the initial table cost for the smaller table, but yes, real value with everything included could be 1 or 2 queries less (based on the initial small table size, which I don't know, just estimated from the sizes).

Reply:
Yes, I meant 14, sorry I wrote too fast :)  Yes, I noticed the upper-bound too.  Regarding 27, really actually around 26 since the long-term storage cost is $0.010/GB, and original table size is listed as 2.6 TB.

Reply:
@tneymanov: variant-optimized tables are generated by default and they are needed for variant lookup queries. Based on this we assume users are required to pay their storage costs.

@pgrosu: if you read the text more carefully you will see we claimed the break point is 14 not 27. I am not sure how you came up with 11, but here the simple calculation: we are looking for smallest `n` that satisfy `n * (2 - 0.005) > 27` which gives us `n > 27 / (2 - 0.005)` or 'n > 13.53'. Please let me know if this makes sense.

Reply:
Ah makes sense, cost of the variant table is non-negotiable, and so, not relevant. Thanks!

Reply:
@samanvp I know, I did the calculation too :)  I wrote 27 instead of 14 as I was focusing on something else.  Below are my calculations:
```
Queries (QP) = $5.00 per TB
Long-term storage (LTS) = $0.010 per GB
Sample-Optimized Table Size (SOTS) = 2.6 TB = 2600 GB
Sample-Optimized Table Storage Price (SOTSP) = LTS * SOTS = $0.01 * 2600 = $26
Sample-Optimized Lookup Query Table Size (SOLQTS) = 1052 MB = 1.052 GB = 0.001052 TB
Variant-Optimized Table Size (VOTS) = 408.13 GB = 0.40813 TB
Variant-Optimized Table Storage Price (VOTSP) = LTS * VOTS = $0.01 * 408.13 = $4.0813‬
Sample-Optimized Query Price (SOQP) = QP * SOLQTS = $5 * 0.001052 = $0.00526
Variant-Optimized Query Price (VOQP) = QP * VOTS = $5 * 0.40813 = $2.04065
Break-Even Rounds (BER) = Number of rounds required to break even

BER * VOQP + VOTSP = BER * SOQP + SOTSP
BER * ( VOQP - SOQP ) = SOTSP - VOTSP
BER = (SOTSP - VOTSP)/(VOQP - SOQP) = 
       = ($26 - $4.0813‬)/($2.04065 - $0.00526) = 
       = $21.9187 / $2.03539 = 10.76879615 ≈ 11
```

Please let me know if this makes sense, or if I misunderstood something.

Thanks,
~p

Reply:
Thanks @pgrosu for detailed explanation.
As @tneymanov  mentioned earlier, storage cost of variant-optimized table is not relevant. If you remove that from your calculations you will get the same answer (14).

Reply:
@samanvp So I assume you mean `$26/($2.04065 - $0.00526) = 12.77396 ≈ 13`, with the assumption the `$4.08` storage cost is placed somewhere else.  Would that be correct?  If so, then it's close enough :)